src/librustc_mir/dataflow/impls/use_def_chain.rs

```diff
@@ -0,0 +1,302 @@
+use std::collections::BTreeMap;
+use std::ops;
+
+use rustc::mir::{self, BasicBlock, Local, Location, Place};
+use rustc::mir::visit::{Visitor, MirVisitable};
+use rustc::util::captures::Captures;
+use rustc_data_structures::bit_set::BitSet;
+use rustc_data_structures::fx::FxHashMap;
+use rustc_data_structures::indexed_vec::{Idx, IndexVec};
+
+use crate::dataflow::{
+    DataflowResults, DataflowResultsCursor,
+    impls::HaveBeenBorrowedLocals,
+    impls::reaching_defs::{DefIndex, DefKind, Definition, ReachingDefinitions, UseVisitor},
+};
+
+type LocationMap<T> = FxHashMap<Location, T>;
+type DefsForLocalDense = IndexVec<Local, BitSet<DefIndex>>;
+type DefsForLocalSparse = BTreeMap<Local, BitSet<DefIndex>>;
+type DefsForUseAtLocation = LocationMap<DefsForLocalSparse>;
```

Comment:
Once there's a pass that uses a `UseDefChain`, these  data structures should be tweaked for efficiency. I think that `DefsForLocalDense` should be a `BitMatrix` and that some uses of `BitSet` can be changed to `HybridBitSet`s.